dropshot/src/router.rs

```diff
@@ -91,21 +91,27 @@ struct HttpRouterNode<Context: ServerContext> {
 enum HttpRouterEdges<Context: ServerContext> {
     /** Outgoing edges for literal paths. */
     Literals(BTreeMap<String, Box<HttpRouterNode<Context>>>),
-    /** Outgoing edges for variable-named paths. */
-    Variable(String, Box<HttpRouterNode<Context>>),
+    /** Outgoing edge for variable-named paths. */
+    VariableSingle(String, Box<HttpRouterNode<Context>>),
+    /** Outgoing edge that consumes all remaining components. */
+    VariableRest(String, Box<HttpRouterNode<Context>>),
 }
 
 /**
  * `PathSegment` represents a segment in a URI path when the router is being
  * configured.  Each segment may be either a literal string or a variable (the
- * latter indicated by being wrapped in braces.
+ * latter indicated by being wrapped in braces). Variables may consume a single
+ * /-delimited segment or several as defined by a regex (currently only `.*` is
```

Comment:
There's a couple of things that make me uncomfortable here but these fears might be specious.

One is that we might be conflating "consumes more than one path segment" with "is matched by a regular expression".  I could imagine someone wanting to use regular expressions to match just a segment, as in `/instances/{instance_id:[0-9]+}/status`.  (I'm not really saying we should try to support that, so much as: if we're trying to leave the door open to regular expressions in path segments, it's not obvious to me now that this _shouldn't_ be possible.)

Another is that I think there's some non-trivial work that goes into correctly separating the path segments in HTTP.   This is the kind of stuff I'm thinking about
https://github.com/oxidecomputer/dropshot/blob/e0201cf97431c4326423998dc575b91fc86970ed/dropshot/src/router.rs#L474-L502
Okay, I mostly punted on that, but the goal was to ensure that there's one place where the correct code would go.  These feel like the sort of things that are fine to get wrong for a while and then really hose you when you run into the case where they're wrong.  I'm a little worried that if we allow variables to match multiple segments, then the consumer will try to interpret this path themselves, and it will become really easy to do something that appears to work in some cases but breaks when the input is percent-encoded or something like that.  (If percent-encoding is a real issue, I could imagine this behavior resulting in a vulnerability where somebody thinks they're rejecting requests that contain `/..`, but they're missing requests that contain `'/%2e%2e'`, and this lets people read /etc/passwd or whatever.  Again, this is potentially already broken today because all I have is a TODO, not a correct implementation, but it would suck for this problem to sprawl.)  Does this go against the principle of "make it hard to do the wrong thing"?

We could potentially fix this by saying that instead of a String, you get an array of path segments or something like that.

Am I being too paranoid?